### PR TITLE
x-live-blog-wrapper

### PR DESCRIPTION
**Pull request by tbergmen**
`x-live-blog-wrapper` is designed to handle initial rendering of live blog posts at the server side and handle live updates at the client side.

**Current features:**
* Render initial posts
* Render hydration data at server side rendering
* Live updates

**Developer notes:**
In next-article, there is a hacky branch that renders this component with data from Wordpress live blogs. This branch is local testing purposes only and is not meant to be merged or used in production.
https://github.com/Financial-Times/next-article/tree/x-live-blog-wrapper-test
https://github.com/Financial-Times/next-article/commit/1ea0698a7b838c1180a8f49c3b901021caaff6c1

### CODE REVIEW

File: components/x-live-blog-wrapper/src/HydratedLiveBlogWrapper.jsx
Comment (apaleslimghost) [inline components/x-live-blog-wrapper/src/HydratedLiveBlogWrapper.jsx:13]: i'd originally intended that using the serialiser would be left for the consuming app, since you might want to reuse the same serialiser for multiple components to reduce the output, and you might have particular requirements for where the `HydrationData` gets output

also i don't know if we need to think coordinating this with the various Page Kit components that output server data; `x-interaction` predates Page Kit and the hydration stuff hasn't been used in production yet

happy to chat further about this! 🙂

File: components/x-live-blog-wrapper/src/LiveEventListener.js
Comment (apaleslimghost) [inline components/x-live-blog-wrapper/src/LiveEventListener.js:40]: please don't dispatch global events from x-dash components. if this component is used in `ft-app`, they way it renders article pages means up to 3 instances of this component could be on the page at once, because of preloaded swipe panes. it looks like you should be able to use `x-interaction` for this use case though?

File: components/x-live-blog-wrapper/package.json
Comment (ker-an) [inline components/x-live-blog-wrapper/package.json:32]: We prob wanna update the node version here or we can remove the field altogether

File: components/x-live-blog-wrapper/src/LiveBlogWrapper.jsx
Comment (simonplend) [inline components/x-live-blog-wrapper/src/LiveBlogWrapper.jsx:30]: I'm wondering if there are pros or cons to creating new arrays when handling each event (as is happening here) vs modifying the existing posts array directly e.g.

- `insertPost` - Add `post` to the beginning of the `posts` array with [`Array.prototype.unshift()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/unshift)
- `updatePost` - Find post with [`Array.prototype.findIndex()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex) and then do `posts[index] = updated`
- `deletePost` - Find post with [`Array.prototype.findIndex()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex) and then use [`Array.prototype.splice()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice) to remove it

The `posts` array could get quite large, so the main benefit I can think of when modifying the array in place is that you're not using more memory by creating copies of the posts array every time we receive a new event. I'm not sure how much this matters though :thinking: Interested to hear your thoughts on this!

File: components/x-live-blog-wrapper/src/LiveBlogWrapper.jsx
Comment (simonplend) [inline components/x-live-blog-wrapper/src/LiveBlogWrapper.jsx:33]: I think it would improve the readability of this statement if it was formatted across multiple lines.

File: components/x-live-blog-wrapper/src/LiveEventListener.js
Comment (simonplend) [inline components/x-live-blog-wrapper/src/LiveEventListener.js:1]: :tada:

File: components/x-live-blog-wrapper/src/HydratedLiveBlogWrapper.jsx
Comment (tbergmen) [reply components/x-live-blog-wrapper/src/HydratedLiveBlogWrapper.jsx:13]: @apaleslimghost that makes sense. Currently, one of the consumer apps (`next-article`) is rendering this component using `x-handlebars`. I couldn't find a way to use serialiser with it. Do you think it is possible?

Would it be acceptable to offer a plain version of this component that doesn't serialise the hydration data as well as having this one as an option?

File: components/x-live-blog-wrapper/src/LiveEventListener.js
Comment (tbergmen) [reply components/x-live-blog-wrapper/src/LiveEventListener.js:40]: The purpose of this event is to notify the consumer app when the state of this component is updated. Would that be possible with `x-interaction`?

File: components/x-live-blog-wrapper/src/LiveBlogWrapper.jsx
Comment (ker-an) [inline components/x-live-blog-wrapper/src/LiveBlogWrapper.jsx:33]: So this `key` here is used by `next-article` to initialise the Origami components?

File: components/x-live-blog-wrapper/src/LiveEventListener.js
Comment (ker-an) [inline components/x-live-blog-wrapper/src/LiveEventListener.js:93]: Good question, I'm not sure. Is the alternative that the topper deals with this?

File: components/x-live-blog-wrapper/src/LiveEventListener.js
Comment (tbergmen) [reply components/x-live-blog-wrapper/src/LiveEventListener.js:40]: To be more specific;

We are listening to server sent events in this component and add new posts (by adding `x-live-blog-post` components to the markup) based on those events. 

`x-live-blog-post` component has Origami components that need to be initialised after rendering the markup. Because we want the consumer app to initialise Origami components, we notify the app using that event when the markup is changed.

File: components/x-live-blog-wrapper/src/LiveBlogWrapper.jsx
Comment (tbergmen) [reply components/x-live-blog-wrapper/src/LiveBlogWrapper.jsx]: @simonplend Thanks for the suggestion. I have created PR for this: https://github.com/Financial-Times/x-dash/pull/517

File: components/x-live-blog-wrapper/src/LiveBlogWrapper.jsx
Comment (tbergmen) [reply components/x-live-blog-wrapper/src/LiveBlogWrapper.jsx]: @ker-an key is a required property by React for items in a list. In this case we are creating a list of LiveBlogPost items and we assign a unique key to each post so that React can identify which posts are changed.

There is more info about this on React documentation: https://reactjs.org/docs/lists-and-keys.html#keys

File: components/x-live-blog-wrapper/src/HydratedLiveBlogWrapper.jsx
Comment (apaleslimghost) [reply components/x-live-blog-wrapper/src/HydratedLiveBlogWrapper.jsx]: sorry, i missed these comments. i think setting `res.locals.serialiser` will work and get passed through to the component. you might need to do `serialiser=@root.serialiser` when rendering it

File: components/x-live-blog-wrapper/src/LiveEventListener.js
Comment (apaleslimghost) [reply components/x-live-blog-wrapper/src/LiveEventListener.js]: ah, okay. you could dispatch the events from the wrapper element and let them bubble instead?

**Review by apaleslimghost — COMMENTED** [review]
overall this looks good, i think if you address my comment about event dispatching it'll be good to go!

File: components/x-live-blog-wrapper/src/LiveEventListener.js
Comment (tbergmen) [reply components/x-live-blog-wrapper/src/LiveEventListener.js]: Thanks @apaleslimghost 

Sorry I forgot to update this thread; we have already updated it to raise the event from the wrapper element: https://github.com/Financial-Times/x-dash/blob/313afe7a016b2911dfabafd6bac251d104832296/components/x-live-blog-wrapper/src/LiveEventListener.js#L52

Do you think it looks ok now?

Comment (apaleslimghost) [reply components/x-live-blog-wrapper/src/LiveEventListener.js]: ah! sorry, i didn't check the current state of this file. yes, that looks good!

**Review by apaleslimghost — APPROVED** [review]
okay, this looks ready! per the [release guidelines](https://github.com/Financial-Times/x-dash/blob/master/docs/components/release-guidelines.md) this should be tested in `ft-app` before merging. this could be as simple as installing it locally and writing a quick spike implementation, and checking that with the Apps team (you might already have done this, idk!)

File: components/x-live-blog-wrapper/src/HydratedLiveBlogWrapper.jsx
Comment (tbergmen) [reply components/x-live-blog-wrapper/src/HydratedLiveBlogWrapper.jsx]: That's good to know. Thanks!

File: components/x-live-blog-wrapper/src/LiveBlogWrapper.jsx
Comment (nickcolley) [inline components/x-live-blog-wrapper/src/LiveBlogWrapper.jsx:7]: When posts are updated and inserted, does the application take charge of informing the user about it?

Comment (GlynnPhillips) [reply components/x-live-blog-wrapper/src/LiveBlogWrapper.jsx:7]: @nickcolley there is currently a notification that appears and takes you to the new post should you want to. This component is new so we are making some of these architectural decisions along the way and one that we have defined yet is how much of that should live in here and how much should live in the application.

The only version of these pages wasn't amazing for accessibility and I don't think it told you about updates or deletes and only mentioned creates. So we must improve this and I think it will be the main factor in "whose" responsibility that is, the app or the component. At this stage I am leaning towards component.

Comment (GlynnPhillips) [reply components/x-live-blog-wrapper/src/LiveBlogWrapper.jsx:7]: Sorry that was a bit of a ramble to acknowledge that yes we need to make sure the user experience is accessible

**Review by apaleslimghost — DISMISSED** [review]
smashing that approve button again since the recent commits dismissed my previous review

**Comment by nickcolley** [comment]
Heya @tbergmen,

Did you get a chance to read this conversation that I had with Glynn?

https://github.com/Financial-Times/x-dash/pull/509#discussion_r478512642

**Review by ker-an — APPROVED** [review]
This is really great work and the comments/documentation really helped me.

**Comment by tbergmen** [comment]
> Heya @tbergmen,
> 
> Did you get a chance to read this conversation that I had with Glynn?
> 
> [#509 (comment)](https://github.com/Financial-Times/x-dash/pull/509#discussion_r478512642)

Hello @nickcolley, yes, accessibility of live updates is in our to-do list. In the first release of live blogs, we won't have live updates. We will revisit this component when we are building that feature.

**Comment by nickcolley** [comment]
@tbergmen thanks for the update, sounds great nice work!

**Pull request merged**
